File: healthcare/healthcare/report/patient_billing/patient_billing.js
// Copyright (c) 2024, earthians Health Informatics Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Patient Billing"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "patient",
            "label": __("Patient"),
            "fieldtype": "Link",
            "options": "Patient",
            "reqd": 1
        },
    ],

    "onload": function(report) {
        // Create a Pie chart after the report is loaded
        let chart_data = {
            labels: ["Help Account", "Paid"],
            datasets: [
                {
                    name: "Status",
                    chartType: "pie",
                    values: [0, 0], // Placeholder values
                }
            ]
        };

        // Create the chart (frappe method to render the chart)
        frappe.chart.create({
            parent: report.page,
            title: "Invoice Status Chart",
            chart_type: "Pie",
            data: chart_data,
            height: 250
        });
    },

    "formatter": function(value, column, data, report) {
        // Highlight status and dynamically update chart
        if (column.fieldname === "status") {
            let chart = report.page.charts[0];  // Access the chart

            // Get current chart data
            let current_data = chart.get_data();

            if (value === "Help Account") {
                // Update Help Account count
                current_data.datasets[0].values[0] += 1;  // Increment Help Account count
                chart.set_data(current_data); // Update chart with new data
                return `<span style="color: red">${value}</span>`;  // Color the status red
            } else if (value === "Paid") {
                // Update Paid count
                current_data.datasets[0].values[1] += 1;  // Increment Paid count
                chart.set_data(current_data); // Update chart with new data
                return `<span style="color: green">${value}</span>`;  // Color the status green
            }
        }
        return value;
    }
};

File: healthcare/healthcare/report/patient_billing/patient_billing.py
import frappe

def execute(filters=None):
    columns, data = get_columns(), get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": "Invoice ID", "fieldname": "invoice_id", "fieldtype": "Link", "options": "Sales Invoice", "width": 200},
        {"label": "Invoice Date", "fieldname": "invoice_date", "fieldtype": "Date", "width": 100},
        {"label": "Patient", "fieldname": "patient", "fieldtype": "Link", "options": "Patient", "width": 150},
        {"label": "Service", "fieldname": "service", "fieldtype": "Data", "width": 300},
        {"label": "Total Amount", "fieldname": "total", "fieldtype": "Currency", "width": 100},
        {"label": "Paid Amount", "fieldname": "net_total", "fieldtype": "Currency", "width": 100},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 100}
    ]

def get_data(filters):
    conditions = []
    if filters.get("from_date") and filters.get("to_date"):
        conditions.append("si.posting_date BETWEEN %(from_date)s AND %(to_date)s")
    if filters.get("customer"):
        conditions.append("si.customer = %(customer)s")
    if filters.get("patient"):
        conditions.append("si.patient = %(patient)s")

    # Join all conditions with AND; if no conditions, defaults to "1=1"
    where_clause = " AND ".join(conditions) if conditions else "1=1"

    query = f"""
        SELECT
            si.name AS invoice_id,
            si.posting_date AS invoice_date,
            si.customer,
            si.patient,
            GROUP_CONCAT(sii.item_name SEPARATOR ', ') AS service,
            si.total AS total,
            si.net_total AS net_total
        FROM
            `tabSales Invoice` si
        LEFT JOIN
            `tabSales Invoice Item` sii ON sii.parent = si.name
        WHERE
            {where_clause}
        GROUP BY
            si.name
        ORDER BY
            si.posting_date DESC
    """
    data = frappe.db.sql(query, filters, as_dict=True)

    # Update status based on the total amount
    for row in data:
        if row["net_total"] <= 1:
            row["status"] = "Help Account"
        else:
            row["status"] = "Paid"

    return data
